### PR TITLE
Marshal Support for Maps

### PR DESCRIPTION
**Pull request by seglberg**
Maps can be marshaled at the top-level or embedded in structs
as exported fields.

---

This is great for dealing with many of the Vici commands where we don't know the key names ahead of time or otherwise would require manually iterating the Message keys and marshaling/unmarshaling sub-fields.

For example with `load-conn`, it allows us to quickly unmarshal into a `map[string]*OurConnStruct` without having to iterate over each key in the Message manually.

It also helps out with our message structures, for example, or IKE Config structs can include simplified Children fields:

```go
type IKEConfig struct {
	Version IKEVersion `json:"version" vici:"version"`

	Encap bool `json:"encap" vici:"encap"`

	DPDDelaySec   int `json:"dpd_delay" vici:"dpd_delay"`
	DPDTimeoutSec int `json:"dpd_timeout" vici:"dpd_timeout"`

	Pools []string `json:"pools" vici:"pools"`

	LocalAuth  []*AuthConfig `json:"local,omitempty"`
	RemoteAuth []*AuthConfig `json:"remote,omitempty"`

	Children map[string]*ChildSAConfig `json:"children" vici:"children"`
}
```

### CODE REVIEW

**Comment by seglberg** [comment]
I missed the fact that `MapRange` is only available since Go 1.12
Looks like the CI is testing with 1.11, so I will find an alternative and re-push.

File: message.go [hunk 1]
```diff
@@ -845,6 +909,43 @@ func (m *Message) unmarshal(v interface{}) error {
...
+	for _, k := range m.keys {
+
+		value, ok := m.data[k]
+		if !ok {
+			continue
+		}
+
+		// Map Value Kind
+		mapElemType := rv.Type().Elem()
+		switch mapElemType.Kind() {
```
Comment (enr0n) [inline message.go]: `rv` is the map here isn't it? So isn't this switch statement always going to follow the `reflect.Map` case?

Shouldn't you be checking the type of `k` instead? And what about a check to make sure `value` is assignable to this map?

File: message.go [hunk 2]
```diff
@@ -915,7 +1016,15 @@ func (m *Message) unmarshalField(field reflect.Value, rv reflect.Value) error {
...
 		}
 
-		return msg.unmarshal(field.Interface())
+		fp := field
+		if field.IsNil() {
+			fp = reflect.New(field.Type().Elem())
+			field.Set(fp)
+		}
+
+		if err := msg.unmarshal(fp.Interface()); err != nil {
```
Comment (enr0n) [inline message.go]: What's the need for the `if` statement here? Why not just `return msg.unmarshal(fp.interface())`?

File: message.go [hunk 3]
```diff
@@ -915,7 +1016,15 @@ func (m *Message) unmarshalField(field reflect.Value, rv reflect.Value) error {
 			return fmt.Errorf("%v: %v", errUnmarshalNonMessage, rv.Type())
 		}
 
-		return msg.unmarshal(field.Interface())
+		fp := field
+		if field.IsNil() {
```
Comment (enr0n) [inline message.go]: Why exactly is this needed? Is this an old issue or due to your changes?

**Comment by enr0n** [comment]
Thanks for another great PR :) The marshaling is going to be really solid after this.

File: message.go (hunk 3 above)
Comment (seglberg) [reply message.go]: Let me find the test case that required this and I will add it to the tests.

File: message.go (hunk 2 above)
Comment (seglberg) [reply message.go]: Not longer needed. I believe I re-ordered something at one point here.

File: message.go (hunk 3 above)
Comment (seglberg) [reply message.go]: I will open a separate request about this. This is for automatically initializing pointers to nested structs.

File: message.go (hunk 1 above)
Comment (seglberg) [reply message.go]: I added a check for the key type.

I also re-wrote this section a bit as it still had a few edge cases it was failing (added new test cases as well).

Basically what is happening is that a value from `MapIndex` cannot be set directly since they are not addressable (also they're Invalid if the key doesn't exist, despite what the docs say about returning a zero value). So we have to use reflect to create a new instance of the map value, set that, and then use `SetMapIndex` to put it into the map.

**Comment by enr0n** [comment]
LGTM.
